tools/image-audit/identity/clustermanager.js

```diff
@@ -0,0 +1,263 @@
+
+    return cluster.id;
+  }
+
+  #add(newCluster) {
+    this.#clusterMap.set(newCluster.id, newCluster);
+    this.#types.add(newCluster.type);
+    return newCluster.id;
+  }
+
+  get(clusterId) {
+    return this.#clusterMap.get(clusterId);
+  }
+
+  reclusterComplete(persistedCluster, consumedCluster) {
+    // This maintains a reference from the old cluster to the new one
+    // so that any async operations automatically pick up this change.
+    this.#clusterMap.set(consumedCluster.id, persistedCluster);
+  }
+
+  identityAdded(identity, cluster) {
+    if (identity.strong) {
+      if (this.#strongIdentityToClusterMap.has(identity.id)) {
+        if (this.#strongIdentityToClusterMap.get(identity.id) !== cluster) {
+          this.#mergeCluster(cluster, identity);
+        }
+      } else {
+        this.#strongIdentityToClusterMap.set(identity.id, cluster);
+      }
+    }
+    // Nothing to do with soft identities
+  }
+
+  identityMoved(identity, updatedOwningClusterId) {
+    if (identity.strong) {
+      this.#strongIdentityToClusterMap.set(identity.id, this.get(updatedOwningClusterId));
+    }
+    // Nothing to do with soft identities
+  }
+
+  identityRemoved(identity, cluster) {
+    if (identity.strong && this.#strongIdentityToClusterMap.get(identity.id) === cluster) {
+      this.#strongIdentityToClusterMap.delete(identity.id);
+    }
+  }
+
+  async detectSimilarity(batchedClusterIds, identityState, concurrency) {
+    const callCount = { value: 0 }; // Use an object to hold the count
+    const clusterInfoWithInstigator = [];
+
+    // Gather cluster info with instigator identities
+    batchedClusterIds.forEach((clusterId) => {
+      const cluster = this.get(clusterId);
+      cluster.identities.forEach((identity) => {
+        if (identity.similarityInstigator) {
+          clusterInfoWithInstigator.push({
+            clusterId: cluster.id,
+            instigatorIdentityType: identity.type,
+            instigatingIdentity: identity,
+          });
+        }
+      });
+    });
+
+    // Initialize identity state if it doesn't exist
+    const similarityIdentityState = identityState[SimilarClusterIdentity.type];
+    if (!similarityIdentityState) {
+      identityState[SimilarClusterIdentity.type] = {};
+    }
+
+    // Process each cluster info sequentially
+    for (const clusterInfo of clusterInfoWithInstigator) {
+      const { clusterId, instigatorIdentityType, instigatingIdentity } = clusterInfo;
+      const instigatingCluster = this.get(clusterId);
+
+      const allClustersWithIdentity = this.getAllWithIdentity(instigatorIdentityType);
+      const clustersToCompare = instigatingIdentity.filterSimilarClusters(allClustersWithIdentity);
+      const { similarityCollaboratorIdentityTypes } = instigatingCluster;
+
+      // Compare clusters sequentially
+      for (const similarCluster of clustersToCompare) {
+        // eslint-disable-next-line no-await-in-loop
+        await this.#compareClustersForSimilarity(
```

Comment:
Does this need to be sequential ? Maybe a `Promise.all` could optimise the speed...?